src/kademlia/dht_tracker.cpp

```diff
@@ -294,26 +251,25 @@ namespace libtorrent { namespace dht
 		, boost::function<void(item const&, bool)> cb
 		, std::string salt)
 	{
-		m_dht.get_item(key, salt, boost::bind(&get_mutable_item_callback, _1, _2, cb));
+		m_dht.get_item(key, salt, cb);
 	}
 
 	void dht_tracker::put_item(entry data
-		, boost::function<void()> cb)
+		, boost::function<void(int)> cb)
 	{
 		std::string flat_data;
 		bencode(std::back_inserter(flat_data), data);
 		sha1_hash target = item_target_id(
 			std::pair<char const*, int>(flat_data.c_str(), flat_data.size()));
 
-		m_dht.get_item(target, boost::bind(&put_immutable_item_callback
-			, _1, cb, data));
+		m_dht.put_item(target, data, cb);
```

Comment:
it's not entirely obvious to me what the benefit of introducing put_item() and get_item() is, over having a get_item() where you can choose to also put once your callback is called. Is there state that's lost that can be more efficiently handed off across traversal_algorithms? It seems like this just makes the interface wider without simplifying the get_item() function. I suppose you can save one or two layers of forwarding boost::function calls.